### ISSUE TITLE
Flag for deterministic GPU operations.

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#3029]
Some GPU operation use AtomicAdd for efficiency when accumulating numbers in a buffer, for summation operations for instance. The problem is that limited-precision floating-point arithmetic is not exactly associative, and AtomicAdd does not guarantee the order in which the operations will be done.
This means that the results are not exactly reproducible numerically.
Ops concerned by that would be:
- Gemm, Gemv, maybe other BLAS Ops: they do not use AtomicAdd for the moment, but could be more efficient if they did
- AdvancedIncSubtensor1_dev20: uses AtomicAdd, no way to disable it, but AdvancedIncSubtensor1 (slow) is deterministic
- CuDNN convolutions: use AtomicAdd for the gradients, no way to disable it, but other convolution implementations do not use them.

We should have a global theano flag to select between the two behaviours, and maybe a finer-grained control at the Op level itself if feasible.

**Comment by nouiz** [comment]
cuBLAS have a way to get some of its implementation faster by AtomicAdd, we only need to provide that interface, we don't need to implement the GPU code ourself.

**Comment by pdoongarwal** [comment]
@nouiz I would like to would like to work on it? How should i gain background for it?

**Comment by nouiz** [comment]
The first step would be to make a deterministic Theano flag in theano/configdefault.py. Then use it in the opt local_gpu_advanced_incsubtensor1 and local_gpua_advanced_incsubtensor to select which implementation (point 2 above) to use.

After that, we need a way to change point 3. For this, in the cudnn op, I would check the that algo must be deterministic if that flag is true.

point 1 is a possible future case to handle if we want the extra speed from it. So nothing to do about it now.

**Comment by gyom** [comment]
Created a pull request to resolve this issue.

https://github.com/Theano/Theano/pull/4209

**Comment by ozancaglayan** [comment]
Just for information,

I tested a GRU based NMT code with and without that `AdvancedIncSubtensor1_dev20` optimization. Here is the mean minibatch update time for 520 updates on a Tesla K40, CUDA 6.5, without CUDNN:

```
In [19]: opt.mean()
Out[19]: 0.8737307692307692

In [20]: noopt.mean()
Out[20]: 0.88592307692307692
```

So I decided to go without the optimization since it doesn't seem to hurt the performance in my setup.

Thanks!

**Comment by nouiz** [comment]
thanks for the information.

**Comment by ozancaglayan** [comment]
I have another comment here.

I run my NMT system 4 times with the default Theano having the non-deterministic behaviour and 2 systems by patching the `sandbox/cuda/opt.py` to always use the `AdvancedIncSubtensor1` op.

BLEU (translation evaluation metric, higher is better) scores for 4 systems with non-deterministic behaviour:
```
35.86
36.13
36.29
35.35
```

The deterministic systems gave a BLEU of `36.88`. This is a pretty significant margin in terms of BLEU score. A BLEU gain of +1 generally deserves a new paper in these days. Since we have lots of data around, I don't think that people trains 5-10 systems to average the scores or to do significance tests. So I'm kindly offering to disable this GPU opt by default to also be in sync with the CPU behaviour. What do you think?

**Comment by nouiz** [comment]
Good point. I'm not sure what to do. In some cases, it gave pretty good
speed up and the non-deterministic behavior wasn't a problem.

I think the good solution is to switch to a default deterministic behavior
for your reason, but make it much faster then the current one. So that the
speed difference isn't as big.

We are working on making a simple Theano flag to easily switch between
those version to make it easy.

I would like to know other people thinking on which default we should use.
I don't know when we will have the time to make a new version. Maybe we
need to change it now until there is a better version.

**Comment by ozancaglayan** [comment]
Do we have an empirical evidence about the slowness of the non-optimized op? As I mentioned above, I see no difference at all in terms of minibatch update times. I can also compare the numbers on a GTX980 + CUDA 7.5 + CuDNNv4 setup if you'd like to.

**Comment by ozancaglayan** [comment]
Hi nouiz,

Do we have any updates regarding this?

**Comment by nouiz** [comment]
It was started here:

https://github.com/Theano/Theano/pull/4212

But with NIPS deadline, I needed to put it aside. But you can use that PR with the flag to just get rid of the non-deterministic. It is just that we don't have all the options we need.

**Comment by nouiz** [comment]
I don't think we have time for the 0.9 release, so I bump that to 0.10 release.

**Comment by wulabs** [comment]
Should it be deterministic across the same architecture/environment but different machines?  Currently with these patches I am able to get same results for successive runs on either GPU or CPU on a single machine but not across different machines even with similar architectures/environments.

**Comment by nouiz** [comment]
The word "similar" is probably your problem.

Small difference of libraries like OpenBLAS vs mkl can change the numerical
results due to how float are reprensented in hardware

https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html

**Comment by wulabs** [comment]
Just to be clear, we tested a number of OSX machines with core i7's on OSX 10.12.1 with nVidia GPUs (tested both CPU and GPU) with the same XCode, Cuda, CuDNN, Python, Keras, NumPy, Theano versions.  Maybe there is a lower level library or two that is not the same.

**Comment by nouiz** [comment]
Did you set numpy random number generator seed? Probably.

Is this in Python 2 or 3? Python 3 have more non-deterministic strucuture
by default.

**Comment by wulabs** [comment]
Yes, we are using np seed at the top before any Keras imports on Python 2.

We were able to achieve the same results across GPU with the latest Xcode8 CLT, Cuda 8.0.55, CUDNN 5105, Theano TOT, Keras TOT, and the theano dnn.conv params set to deterministic. For NumPy and all other Python frameworks we use the default from pip. The machines are the same architecture (Core i7, nvidia 650/750 GT) on OSX 10.12.2.

On CPU we are able to achieve same results across CPU with theano optimizer=None. It seems the Keras MaxPool layer is what caused the divergence.

**Comment by nouiz** [comment]
optimizer=None is a very bad practice. You should never use it. Try with optimizer=fast_compile.

**Comment by nouiz** [comment]
Update, now there is the flag deterministic=more that will do what the diff in this issue does.
